Dockerfile

```diff
@@ -245,32 +249,31 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 
 # dockercli
 FROM base AS dockercli-src
-WORKDIR /tmp/dockercli
-RUN git init . && git remote add origin "https://github.com/docker/cli.git"
-ARG DOCKERCLI_VERSION
-RUN git fetch -q --depth 1 origin "${DOCKERCLI_VERSION}" +refs/tags/*:refs/tags/* && git checkout -q FETCH_HEAD
-RUN [ -d ./components/cli ] && mv ./components/cli /usr/src/dockercli || mv /tmp/dockercli /usr/src/dockercli
 WORKDIR /usr/src/dockercli
+COPY hack/dockerfile/cli.sh ./build.sh
+ARG DOCKERCLI_REPOSITORY
+RUN git init . && git remote add origin ${DOCKERCLI_REPOSITORY}
+ARG DOCKERCLI_VERSION
+ARG DOCKERCLI_INTEGRATION_VERSION
+RUN git fetch -q --depth 1 origin "${DOCKERCLI_INTEGRATION_VERSION}" "${DOCKERCLI_VERSION}" +refs/tags/*:refs/tags/*
 
 FROM base AS dockercli
 WORKDIR /go/src/github.com/docker/cli
 ARG DOCKERCLI_VERSION
-ARG DOCKERCLI_CHANNEL=stable
 ARG TARGETPLATFORM
 RUN xx-apt-get install -y --no-install-recommends gcc libc6-dev
 RUN --mount=from=dockercli-src,src=/usr/src/dockercli,rw \
-    --mount=type=cache,target=/root/.cache/go-build,id=dockercli-build-$TARGETPLATFORM <<EOT
-  set -e
-  DOWNLOAD_URL="https://download.docker.com/linux/static/${DOCKERCLI_CHANNEL}/$(xx-info march)/docker-${DOCKERCLI_VERSION#v}.tgz"
-  if curl --head --silent --fail "${DOWNLOAD_URL}" 1>/dev/null 2>&1; then
-    mkdir /build
-    curl -Ls "${DOWNLOAD_URL}" | tar -xz docker/docker
-    mv docker/docker /build/docker
-  else
-    CGO_ENABLED=0 xx-go build -o /build/docker ./cmd/docker
-  fi
-  xx-verify /build/docker
-EOT
+    --mount=type=cache,target=/root/.cache/go-build,id=dockercli-build-$TARGETPLATFORM \
+    ./build.sh ${DOCKERCLI_VERSION} /build
+
+FROM base AS dockercli-integration
+WORKDIR /go/src/github.com/docker/cli
+ARG DOCKERCLI_INTEGRATION_VERSION
+ARG TARGETPLATFORM
+RUN xx-apt-get install -y --no-install-recommends gcc libc6-dev
```

Comment:
Should the `ARG`s be moved below this `RUN` so as not to invalidate the build cache? Or is BuildKit smart enough that it doesn't matter?

Why does gcc need to be installed given that `cli.sh` sets `CGO_ENABLED=0`?

Reply:
Yeah, it should be below the `RUN` as changing the arg will invalidate the cache.

As for the `gcc` - not entirely sure. I just left it as it's what it's been doing before my changes.
I just tested it and seems to also work without it (for both `go build` and `./scripts/build/binary`).

The `gcc + libc` installation was added here https://github.com/moby/moby/commit/a4bc2922c24c0d6a92a817847dff5afb55f75ba3. @crazy-max - do you remember why it was needed?

Reply:
Indeed does not look necessary